Unify shapes of ActiveModel::Attributes

casperisfine opened the pull request:
`if defined?(@ivar)` a performance anti-pattern in Ruby 3.2+ even more so with YJIT.

This pattern cause the object shape to be inconsistent which slow downs instance variable access.

```
ruby 3.2.1 (2023-02-08 revision 31819e82c8) [arm64-darwin22]
       #value (orig):  1811240.2 i/s
        #value (opt):  2045238.9 i/s - 1.13x  faster
```

```
ruby 3.2.1 (2023-02-08 revision 31819e82c8) +YJIT [arm64-darwin22]
       #value (orig):  4379180.3 i/s
        #value (opt):  6347280.7 i/s - 1.45x  faster
```

Benchmark: https://gist.github.com/casperisfine/872f0a486b5ccdf90d9feb830c76d9ad

### Backward compatibility

There is a big backward compatibility concern here, and I'm not sure we can actually make this change safety.

Until https://github.com/rails/rails/pull/47747, when you serialized an `ActiveRecord::Base` instance, lots of `ActiveModel::Attribute` instances would be serialized with it. Which means this optimization would break any instance serialized with an older Rails.

Currently the PR is only enabling the optimization if you switched to the new AR Marshal format, but:

  - You can switch to that format and still deserialize old payloads, this PR would break that.
  - Active Record may not be the only way these are serialized. We may break more stuff.

I can't think of any decent way to do this optimization while still retaining perfect backward/forward compatibility...

cc @tenderlove, I wonder if you have opinions here.

byroot committed:
Unify shapes of ActiveModel::Attributes

`if defined?(@ivar)` a performance anti-pattern in Ruby 3.2+
even more so with YJIT.

This pattern cause the object shape to be inconsistent which slow
downs instance variable access.

```
ruby 3.2.1 (2023-02-08 revision 31819e82c8) [arm64-darwin22]
       #value (orig):  1811240.2 i/s
        #value (opt):  2045238.9 i/s - 1.13x  faster
```

```
ruby 3.2.1 (2023-02-08 revision 31819e82c8) +YJIT [arm64-darwin22]
       #value (orig):  4379180.3 i/s
        #value (opt):  6347280.7 i/s - 1.45x  faster
```

Benchmark: https://gist.github.com/casperisfine/872f0a486b5ccdf90d9feb830c76d9ad

rafaelfranca commented on activemodel/lib/active_model/attribute.rb:
```diff
@@ -4,6 +4,12 @@
+    # We can't use an anonymous object token here, because Attribute instances may be
+    # serialized with Marshal, so it would restore a different object.
+    # We use a very specific symbol instead.
+    UNDEF = :__ActiveModel__Attribute__UNDEF__
```
```suggestion
    UNDEF = :__ActiveModel__Attribute__UNDEF__ # :nodoc:
```

rafaelfranca commented on activemodel/lib/active_model/attribute.rb:
```diff
@@ -24,24 +30,122 @@ def null(name)
-      @type = type
-      @original_attribute = original_attribute
-      @value = value unless value.nil?
+    module LegacyMethods
```
```suggestion
    module LegacyMethods # :nodoc:
```

rafaelfranca commented on activemodel/lib/active_model/attribute.rb:
```diff
@@ -24,24 +30,122 @@ def null(name)
-      @value
+    include LegacyMethods
+
+    module ModernMethods
```
```suggestion
    module ModernMethods # :nodoc:
```

rafaelfranca commented:
I don't remember but it is possible to define custom marshal loading to keep backwards compatibility?

jonathanhefner reviewed (commented):
> Currently the PR is only enabling the optimization if you switched to the new AR Marshal format, but:
> 
> * You can switch to that format and still deserialize old payloads, this PR would break that.
> * Active Record may not be the only way these are serialized. We may break more stuff.

What about using separate boolean ivars to indicate undefined-ness?  Then I think we could unify the implementations with something like:

```ruby
attr_reader :has_value, :has_value_for_database

def initialize(name, value_before_type_cast, type, original_attribute = nil, value = nil)
  @name = name
  @value_before_type_cast = value_before_type_cast
  @type = type
  @original_attribute = original_attribute
  @value = value
  @has_value = !value.nil?
  @value_for_database = nil
  @has_value_for_database = false
end

def has_value?
  has_value || (has_value.nil? && defined?(@value))
  # Or possibly:
  # has_value || (@has_value = !!defined?(@value) if has_value.nil?)
end
alias :has_been_read? :has_value?

def has_value_for_database?
  has_value_for_database || (has_value_for_database.nil? && defined?(@value_for_database))
end
```

Since old serializations will not have the boolean ivars, we can detect them with `has_value.nil?` and `has_value_for_database.nil?`, and handle them by falling back to `defined?`.

casperisfine commented:
> I don't remember but it is possible to define custom marshal loading to keep backwards compatibility?

Yes and no. Payload that were serialized form versions that  didn't have a `marshal_dump` method, won't invoke `marshal_load`.

So if we want to keep compat, we need to handle instance that were serialized in 7.0 :/

casperisfine commented:
> Since old serializations will not have the boolean ivars, we can detect them with `has_value.nil?` and `has_value_for_database.nil?`, and handle them by falling back to `defined?`.

Yes, except to avoid warnings in 2.7, we'll also have to check for `defined?`, I can try, but I fear we might lose a good part of the optimization -_-

casperisfine commented:
Also it's not so much forward compatibility that is my problem, but backward, as in `Marshal.dump(rails_7_1_object)` should be loadable from `7.0` so that you can do a rolling release of `7.1`.

Ideally when you first ship 7.1 without the new defaults enabled, it should generate exactly the same payloads as 7.0 used to.

casperisfine commented:
Hum, in my mind the contract was that we do keep YAML compact for long time, but Marshal is only across one version to the next. Did I imagine this?

rafaelfranca commented:
oh yeah. I forgot Marshal is not used in database, so I'd not worry with backwards compatibly. If the new behavior is enabled, we can just tell people to bump the cache version.

jonathanhefner commented:
> Yes, except to avoid warnings in 2.7, we'll also have to check for `defined?`

In which cases?  I was thinking we could replace all occurrences of `defined?(@value)` with `has_value?`, and `attr_reader :has_value` would eliminate the need for `defined?(@has_value)`.  (Likewise for `@value_for_database` / `has_value_for_database`.)

> Also it's not so much forward compatibility that is my problem, but backward, as in `Marshal.dump(rails_7_1_object)` should be loadable from `7.0` so that you can do a rolling release of `7.1`.

Could that be handled by a `config.*` value + conditional definition of `initialize`?

casperisfine commented:
> In which cases?

If you load an object that was serialized by `7.0` from `7.1`, `@has_value` won't be defined.

>  attr_reader :has_value would eliminate the need for defined?(@has_value)

Hum, that's a good point worth testing. 

> Could that be handled by a config.* value + conditional definition of initialize?

That's already the case in this PR. I piggy back on the config introduced in https://github.com/rails/rails/pull/47747

tenderlove commented:
One other thing that may (or may not) be important.  `defined?(@iv)` didn't have an inline cache in Ruby 3.2, [but it will in 3.3](https://github.com/ruby/ruby/pull/7433), and I think YJIT will have support (I paired on support but I don't know if it's been upstreamed yet).

Obviously we should reduce the number of shapes in the system, but maybe the performance impact isn't as great in Ruby 3.3?

casperisfine commented:
> but maybe the performance impact isn't as great in Ruby 3.3?

Right, but it's still basically a year away, so I'd like to reap some benefits now if possible.

casperisfine commented:
Ok, so one solution could be to keep that code unchanged, and to use another class entirely e.g. `Attribute71`. So we can have both active at the same time.

But that's quite ugly, so I'm also tempted to just delay this post 7.1 release, and just remove the old format then.

byroot committed:
Active Model Attribute: Use a secondary flag instead of a sentinel value

The `UNDEF.equal?` overheadis significant enough that what is gained
by higher cache rate is lost doing the check.

It's because we have to both lookup the constant, and then call
`equal?` which surprisingly doesn't have an optimized opcode in the
interpreter. But it does in YJIT, hence why it's still faster with
it enabled.

So instead we can use a secondary instance variable to keep a flag.

Prior to this Attribute instances had 6 ivars, now they have 8, so
they still fit in the same size 80 slots, meaning the memory usage
is unchanged.
